File: Go/lzw12.go
// Bradford Arrington 2025
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
)

const (
	BITS          = 12
	MAX_CODE      = (1 << BITS) - 1
	TABLE_SIZE    = 5021
	END_OF_STREAM = 256
	FIRST_CODE    = 257
	UNUSED        = -1
)

var CompressionName = "LZW 12 Bit Encoder"
var Usage = "in-file out-file\n\n"

type DictionaryEntry struct {
	CodeValue  int
	ParentCode int
	Character  byte
}

var dict [TABLE_SIZE]DictionaryEntry
var decodeStack [TABLE_SIZE]byte

// FindChildNode finds the index of a child node in the dictionary
func FindChildNode(parentCode int, childCharacter int) uint {
	index := (int(childCharacter) << (BITS - 8)) ^ parentCode
	offset := 1
	if index != 0 {
		offset = TABLE_SIZE - index
	}

	for {
		if dict[index].CodeValue == UNUSED {
			return uint(index)
		}

		if dict[index].ParentCode == parentCode && dict[index].Character == byte(childCharacter) {
			return uint(index)
		}

		index -= offset
		if index < 0 {
			index += TABLE_SIZE
		}
	}
}

func DecodeString(count uint32, code uint32) uint32 {
	currentCode := int(code)
	for currentCode > 255 {
		decodeStack[count] = dict[currentCode].Character
		count++
		currentCode = dict[currentCode].ParentCode
	}
	decodeStack[count] = byte(currentCode)
	count++
	return count
}

// readByte returns the next byte from the file as int,
// or -1 if EOF is reached.
func readByte(reader *bufio.Reader) int {
	b, err := reader.ReadByte()
	if err == io.EOF {
		return -1
	}
	if err != nil {
		fmt.Println("Error:", err)
		return -1
	}
	return int(b)
}

func CompressFile(input *os.File, output *BIT_FILE, argc int, argv []string) error {
	nextCode := FIRST_CODE
	var character int
	var stringCode int
	var index uint
	var i int

	for i = 0; i < TABLE_SIZE; i++ {
		dict[i].CodeValue = UNUSED
	}

	reader := bufio.NewReader(input)
	stringCode = readByte(reader)
	if stringCode == -1 {
		stringCode = END_OF_STREAM
	}

	for {
		character = readByte(reader)
		if character == -1 {
			break
		}

		index = FindChildNode(stringCode, character)

		if dict[index].CodeValue != -1 {
			stringCode = dict[index].CodeValue
		} else {
			if nextCode <= MAX_CODE {
				dict[index].CodeValue = nextCode
				dict[index].ParentCode = stringCode
				dict[index].Character = byte(character)
				nextCode++
			}
			OutputBits(output, uint32(stringCode), BITS)
			stringCode = character
		}
	}
	OutputBits(output, uint32(stringCode), BITS)
	OutputBits(output, uint32(END_OF_STREAM), BITS)

	for i = 0; i < argc; i++ {
		fmt.Printf("Unknown argument: %s\n", argv[i])
	}
	return nil
}

func ExpandFile(input *BIT_FILE, output *os.File, argc int, argv []string) error {
	nextCode := FIRST_CODE
	var newCode uint32
	var oldCode uint32
	var character int
	var count uint32

	// Read the first code
	firstCode, err := InputBits(input, BITS)
	if err != nil {
		if err != io.EOF {
			return fmt.Errorf("error reading first code: %w", err)
		}
		return nil // Empty input
	}
	oldCode = firstCode
	if oldCode == END_OF_STREAM {
		return nil
	}

	character = int(oldCode)
	_, err = output.Write([]byte{byte(oldCode)})
	if err != nil {
		return fmt.Errorf("error writing first byte: %w", err)
	}

	// Process input
	for {
		nextCodeInput, err := InputBits(input, BITS)
		if err != nil {
			if err != io.EOF {
				return fmt.Errorf("error reading code: %w", err)
			}
			break
		}
		newCode = nextCodeInput
		if newCode == END_OF_STREAM {
			break
		}

		if newCode >= uint32(nextCode) {
			decodeStack[0] = byte(character)
			count = DecodeString(1, oldCode)
		} else {
			count = DecodeString(0, newCode)
		}

		character = int(decodeStack[count-1])

		// Write decoded string to output
		for count > 0 {
			count--
			_, err := output.Write([]byte{decodeStack[count]})
			if err != nil {
				return fmt.Errorf("error writing decoded byte: %w", err)
			}
		}

		if nextCode <= MAX_CODE {
			dict[nextCode].ParentCode = int(oldCode)
			dict[nextCode].Character = byte(character)
			nextCode++
		}

		oldCode = newCode
	}

	for i := 0; i < argc; i++ {
		fmt.Printf("Unknown argument: %s\n", argv[len(argv)-argc+i])
	}

	return nil
}

File: Lzw12.cs
// Bradford Arrington 2025
using System;
using System.IO;
using System.IO.Compression;

partial class Compressor
{
    public string CompressionName = "LZW 12 Bit Encoder";
    public static string Usage = "in-file out-file\n\n";
    private const int BITS = 12;
    private const int MAX_CODE = (1 << BITS) - 1;
    private const int TABLE_SIZE = 5021;
    private const int END_OF_STREAM = 256;
    private const int FIRST_CODE = 257;
    private const int UNUSED = -1;

    private struct DictionaryEntry
    {
        public int CodeValue;
        public int ParentCode;
        public char Character;
    }

    private static DictionaryEntry[] dict = new DictionaryEntry[TABLE_SIZE];
    private static char[] decodeStack = new char[TABLE_SIZE];
    public static uint FindChildNode(int parentCode, int childCharacter)
    {
        int index = (childCharacter << (BITS - 8)) ^ parentCode;
        int offset = index == 0 ? 1 : TABLE_SIZE - index;

        while (true)
        {
            if (dict[index].CodeValue == UNUSED)
            {
                return (uint)index;
            }

            if (dict[index].ParentCode == parentCode && dict[index].Character == (char)childCharacter)
            {
                return (uint)index;
            }

            index -= offset;
            if (index < 0)
            {
                index += TABLE_SIZE;
            }
        }
    }

    private static uint DecodeString(uint count, uint code)
    {
        while (code > 255)
        {
            decodeStack[count++] = dict[code].Character;
            code = (uint)dict[code].ParentCode;
        }

        decodeStack[count++] = (char)code;
        return count;
    }
    public void CompressFile(Stream input, BitFile output, int argc, string[] argv)
    {
        int nextCode = FIRST_CODE;
        int character;
        int stringCode;
        uint index;

        // Initialize dictionary
        for (int i = 0; i < TABLE_SIZE; i++)
        {
            dict[i].CodeValue = UNUSED;
        }

        // Read the first character
        stringCode = input.ReadByte();
        if (stringCode == -1) // EOF
        {
            stringCode = END_OF_STREAM;
        }

        // Process input
        while ((character = input.ReadByte()) != -1) // EOF
        {
            index = FindChildNode(stringCode, character);

            if (dict[index].CodeValue != UNUSED)
            {
                stringCode = dict[index].CodeValue;
            }
            else
            {
                if (nextCode <= MAX_CODE)
                {
                    dict[index].CodeValue = nextCode++;
                    dict[index].ParentCode = stringCode;
                    dict[index].Character = (char)character;
                }

                output.OutputBits((uint)stringCode, BITS);
                stringCode = character;
            }
        }

        // Write the last string and end-of-stream marker
        output.OutputBits((uint)stringCode, BITS);
        output.OutputBits((uint)END_OF_STREAM, BITS);

        while (argc-- > 0)
            Console.WriteLine($"Unknown argument: {argv[argv.Length - argc - 1]}");
    }

    public void ExpandFile(BitFile input, Stream output, int argc, string[] argv)
    {
        uint nextCode = FIRST_CODE;
        uint newCode;
        uint oldCode;
        int character;
        uint count;

        // Read the first code
        oldCode = (uint)input.InputBits(BITS);
        if (oldCode == END_OF_STREAM)
        {
            return;
        }

        character = (int)oldCode;
        output.WriteByte((byte)oldCode);

        // Process input
        while ((newCode = (uint)input.InputBits(BITS)) != END_OF_STREAM)
        {
            if (newCode >= nextCode)
            {
                decodeStack[0] = (char)character;
                count = DecodeString(1, oldCode);
            }
            else
            {
                count = DecodeString(0, newCode);
            }

            character = decodeStack[count - 1];

            // Write decoded string to output
            while (count > 0)
            {
                output.WriteByte((byte)decodeStack[--count]);
            }

            if (nextCode <= MAX_CODE)
            {
                dict[nextCode].ParentCode = (int)oldCode;
                dict[nextCode].Character = (char)character;
                nextCode++;
            }

            oldCode = newCode;
        }

        while (argc-- > 0)
            Console.WriteLine($"Unknown argument: {argv[argv.Length - argc - 1]}");
    }
}
